### DISCUSSION TITLE
Noob question: how to manually free the memory of an array

### DISCUSSION THREAD

**Discussion by TsXor** [pyopencl/pyopencl#633] [Q&A] [answered]
For example, I want to cauculate:
`m = n**2 + p**2`
When I execute this line of code, I can see the GPU memory usage increasing (because a new array is assigned to m, right?).
And when I want to free the memory used by `n` and `p`, what should I do? I tried del, but it won't help.

Can ary.data.release free the memory used by ary?

**Comment by TsXor** [comment]
For another example:
```
a = somewhatarray
b = somewhatarray
x = a**2
x = b**2
```
Look at the task manager, 4 increase in memory comsumption can be observed (by adding breakpoint() between every line), but no decrease.
Expected: after `x = b**2`, the memory for a**2 should be freed.

**Comment by inducer** [comment, marked as answer]
Memory will be freed after the last reference to an object goes away. At `x = b**2`, the local variables `a` and `b` are still holding references to the input arrays.

To manually free memory (technically: release the reference to the buffer, because OpenCL has its own refcounting): https://documen.tician.de/pyopencl/runtime_memory.html#pyopencl.MemoryObject.release

i.e.

```
a.base_data.release()
```

**Reply by TsXor** [reply]
However, I think a\*\*2 is not a, and it is just a thing returned by the magic \_\_pow\_\_, and x is the only reference to it.

**Reply by TsXor** [reply]
If a\*\*2 is sonewhat a, then memory usage shouldn't increase (at least visibly) when x=a\*\*2 is executed.
